### PR TITLE
[ENH] Add LeJEPA core building blocks

### PR DESCRIPTION
**Pull request by RecreationalMath**
Related to #1888 (PR 1 of 3).

## Description

First of three PRs to complete the LeJEPA implementation per [task breakdown](https://github.com/lightly-ai/lightly/issues/1888#issuecomment-3996087102).
This PR addresses "Section 1) Core LeJEPA Building Blocks" from that breakdown.

#### Core LeJEPA Building Blocks

- [x] Add `lightly/models/modules/lejepa.py` with reusable LeJEPA components:
    - [x] `LeJEPAEncoder` (wrapper around backbone + projector forward)
    - [x] `lejepa_invariance_loss(proj: Tensor) -> Tensor` implementing `(proj.mean(0) - proj).square().mean()`
    - [x] `LeJEPALoss` (combines SIGReg + invariance with lambda)
- [x] Add a dedicated projector head in `lightly/models/modules/heads.py`:
    - [x] `LeJEPAProjectionHead` (MLP style consistent with other methods, e.g. BN/ReLU blocks)
- [x] Export new classes/functions:
    - [x] Update `lightly/models/modules/__init__.py` to expose LeJEPA module + projector. Exposes `LeJEPAEncoder`, `LeJEPALoss`, and `LeJEPAProjectionHead`. `lejepa_invariance_loss` remains importable from the module directly, following the same convention as VICReg's standalone `invariance_loss` / `variance_loss` / `covariance_loss` helpers (not exported at the package level).
- [x] Add/extend tests together with this PR:
    - [x] projector forward/backward shape test. Forward shape is covered via the existing `test_single_projection_head` sweep in `tests/models/test_ProjectionHeads.py`; backward shape is covered by the new `test_lejepa_projection_head_backward` in the same file.
    - [x] combined LeJEPA loss wiring test (SIGReg + inv + lambda). Covered by `test_lambda_zero_equals_pure_invariance` and `test_lambda_one_equals_pure_sigreg` in `tests/models/modules/test_lejepa.py::TestLeJEPALoss`.

All default values follow the [reference implementation](https://github.com/galilai-group/lejepa) and are configurable via constructor args.

PRs 2 and 3 (transforms, examples/docs) will follow.

## Tests

- [x] My change is covered by existing tests.
- [x] My change needs new tests.
- [x] I have added/adapted the tests accordingly.
- [x] I have manually tested the change.

New tests live in `tests/models/modules/test_lejepa.py` (invariance + loss + encoder classes, 10 test methods) and `tests/models/test_ProjectionHeads.py` (`LeJEPAProjectionHead` added to the shared sweep, plus `test_lejepa_projection_head_backward`).

## Documentation

- [x] I have added docstrings to all public functions/methods.
- [x] My change requires a change to the documentation (`.rst` files).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by RecreationalMath** [comment]
Questions for reviewer:

1. Misleading filename `lightly/loss/lejepa_loss.py`
Per IgorSusmelj's task layout `LeJEPALoss` and `lejepa_invariance_loss` are added in `lightly/models/modules/lejepa.py`, leaving `lightly/loss/lejepa_loss.py` to contain only `SIGReg`. This filename is misleading.
The package-level import `from lightly.loss import SIGReg` (shipped in v1.15.23) would survive a rename via the `__init__.py` re-export, but the deeper submodule path `from lightly.loss.lejepa_loss import SIGReg` would break without a shim if any external code/user is already using it. Let me know what you think, should I rename the file in a follow-up PR?

2. `LeJEPALoss` not listed in `lightly.loss.rst`
`LeJEPALoss` is documented via the `.lejepa` automodule section in `lightly.models.rst`, not in `lightly.loss.rst`. Users looking up the LeJEPA loss in the API reference will check `lightly.loss` first and not find it, which hurts discoverability.
Adding a `.. seealso::` or `.. autoclass::` cross-reference in `lightly.loss.rst` that points to `lightly.models.modules.lejepa.LeJEPALoss` would close this gap without duplicating the docstring. Let me know what you think, should I add the cross-reference?

3. Loss tests under `tests/models/modules/`
`TestLeJEPAInvarianceLoss` and `TestLeJEPALoss` live in `tests/models/modules/test_lejepa.py`, mirroring the new source location under `lightly/models/modules/lejepa.py`. `tests/loss/test_lejepa_sigreg_loss.py` now contains only `TestSIGReg`.
Reviewers or contributors scanning for "LeJEPA loss tests" will check `tests/loss/` first and not find them there. Though keeping tests co-located with their source is the more consistent option. Let me know what you think, should the loss tests move to `tests/loss/`?

**Comment by RecreationalMath** [comment]
@IgorSusmelj @yutong-xiang-97 - could you please review this PR? Thanks!

**Review by gabrielfruet — COMMENTED** [review]
Regarding your questions, a rule of thumb that I would say is, everything that is loss should go into the `lightly.loss.lejepa_loss`  and everything that is model should go into `lightly.model.modules.lejepa` (e.g encoders) 

You should also put the tests in the correct place given this modification. 

Loss tests goes into `tests/loss/test_lejepa_loss.py`
Loss goes into `lightly/loss/lejepa_loss.py` (e.g invariance loss)

Model goes into `lightly/models/modules/lejepa.py`
Model tests goes into `tests/models/modules/test_lejepa.py`

File: tests/models/modules/test_lejepa.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        torch.manual_seed(42)
+        lejepa_loss = lejepa_fn(proj)
+
+        torch.manual_seed(42)
+        sigreg_loss = sigreg_fn(proj)
+
+        assert torch.allclose(lejepa_loss, sigreg_loss)
+
+
+class TestLeJEPAEncoder:
```
Comment (gabrielfruet) [inline tests/models/modules/test_lejepa.py:18]: https://github.com/RecreationalMath/lightly/blob/929b9d6c6a7816047dee212497a3f25e8c95eaa9/tests/models/modules/test_ijepa_timm.py#L17-L71

You can use this test suite as reference. It's good that you tested the backward pass. Should be also interesting to test the cuda forward

File: tests/models/modules/test_lejepa.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    def test_backward(self) -> None:
+        torch.manual_seed(0)
+        proj = torch.randn(8, 32, 128, requires_grad=True)
+        loss = lejepa_invariance_loss(proj)
+        loss.backward()
+        assert proj.grad is not None
+        assert proj.grad.shape == proj.shape
+
+
+class TestLeJEPALoss:
```
Comment (gabrielfruet) [inline tests/models/modules/test_lejepa.py]: Move to the `tests/loss/test_lejepa_loss.py`

File: tests/models/modules/test_lejepa.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+def _make_encoder() -> LeJEPAEncoder:
+    backbone = nn.Sequential(
+        nn.Conv2d(3, 8, kernel_size=3, padding=1),
+        nn.AdaptiveAvgPool2d(1),
+    )
+    projection_head = LeJEPAProjectionHead(input_dim=8, hidden_dim=16, output_dim=4)
+    return LeJEPAEncoder(backbone=backbone, projection_head=projection_head)
+
+
+class TestLeJEPAInvarianceLoss:
```
Comment (gabrielfruet) [inline tests/models/modules/test_lejepa.py]: Move to `tests/loss/test_lejepa_loss.py`

File: lightly/models/modules/lejepa.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            number of views, ``N`` is the batch size, and ``D`` is the
+            projection dimensionality.
+
+    Returns:
+        Scalar invariance loss.
+    """
+    return (proj.mean(0) - proj).square().mean()
+
+
+class LeJEPALoss(nn.Module):
```
Comment (gabrielfruet) [inline lightly/models/modules/lejepa.py]: Move to `lightly/loss/lejepa_loss`. Also dont forget to update the `docs/source/lightly.loss.rst`

File: lightly/models/modules/lejepa.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
+"""Reusable LeJEPA components: invariance loss, combined loss, and encoder."""
+
+import torch
+from torch import Tensor, nn
+
+from lightly.loss.lejepa_loss import SIGReg
+from lightly.models.modules.heads import LeJEPAProjectionHead
+
+
+def lejepa_invariance_loss(proj: torch.Tensor) -> torch.Tensor:
```
Comment (gabrielfruet) [inline lightly/models/modules/lejepa.py]: Move to `lightly/loss/lejepa_loss`. Also dont forget to update the `docs/source/lightly.loss.rst`

**Review by gabrielfruet — CHANGES_REQUESTED** [review]

**Review by gabrielfruet — COMMENTED** [review]
Work is very nice! The implementation is very good. Sorry for the misalignment, but you should stick to what I said in an earlier comment in this PR (where things should live).

If you have any question, please feel free to reach out!

File: lightly/models/modules/heads.py [hunk 6]
```diff
@@ -776,6 +776,50 @@ def forward(self, x: Tensor) -> Tensor:
...
+        hidden_layers = [
+            (hidden_dim, hidden_dim, nn.BatchNorm1d(hidden_dim), nn.ReLU())
+            for _ in range(num_layers - 2)  # Exclude first and last layer.
+        ]
+        super().__init__(
+            [
+                (input_dim, hidden_dim, nn.BatchNorm1d(hidden_dim), nn.ReLU()),
+                *hidden_layers,
+                (hidden_dim, output_dim, None, None),
+            ]
```
[the remark this answers is not in the document]
Comment (gabrielfruet) [reply lightly/models/modules/heads.py:822]: Raising a meaningful error is useful

**Review by gabrielfruet — APPROVED** [review]

**Comment by RecreationalMath** [comment]
Thanks for the review and for pushing the structural reorg, docs update and validation directly @gabrielfruet! 

Heads-up: `TestSIGReg` ended up duplicated between `tests/loss/test_lejepa_sigreg_loss.py` and the new `tests/loss/test_lejepa_loss.py`. Looks like this happened because of a missed rebase between the merging of #1923 and #1916. I'll open a small follow-up PR to address this. Once done with that, I will pick up PR 2 (transforms) and PR 3 (examples + docs) next, as per the lejepa plan.

**Comment by gabrielfruet** [comment]
@RecreationalMath I would appreciate that! 

I'm already implementing the transforms + examples. I'll need to make some modifications to the loss because we are not supporting the global x local views from the paper. I think we implemented inspired by the minimal implementation and missed the global x local part. 

I think today or tomorrow I'll open a PR for that. Would be nice if you could take a look also. 

But feel free to open a PR to fix the duplicated tests.

**Comment by RecreationalMath** [comment]
> @RecreationalMath I would appreciate that!
> 
> I'm already implementing the transforms + examples. I'll need to make some modifications to the loss because we are not supporting the global x local views from the paper. I think we implemented inspired by the minimal implementation and missed the global x local part.
> 
> I think today or tomorrow I'll open a PR for that. Would be nice if you could take a look also.

Sure. Please tag me on the PR once it's ready for review.

> But feel free to open a PR to fix the duplicated tests.

~~Will do.~~ Please review it at #1925 @gabrielfruet
